### PR TITLE
Develop > Master release

### PR DESCRIPTION
**Pull request by ericbutz**

### CODE REVIEW

File: contracts/fio.address/fio.address.cpp [hunk 1]
```diff
@@ -523,26 +525,36 @@ namespace fioio {
...
+                            wasFound = true;
+                            break;
+                        }
                     }
                 }
+
                 if(!wasFound){
-                    fio_400_assert(fioname_iter->addresses.size() != 100, "token_code", tpa->token_code, "Maximum token codes mapped to single FIO Address reached. Only 100 can be mapped.",
-                                   ErrorInvalidFioNameFormat);
+                    fio_400_assert(fioname_iter->addresses.size() != MAX_SET_ADDRESSES, "token_code", tpa->token_code, "Maximum token codes mapped to single FIO Address reached. Only 200 can be mapped.",
```
Comment (lukestokes) [inline contracts/fio.address/fio.address.cpp:556]: Can we use MAX_SET_ADDRESSES in the error message itself?

**Review by fiocasey — APPROVED** [review]

File: contracts/fio.address/fio.address.cpp [hunk 2]
```diff
@@ -1612,6 +1637,82 @@ namespace fioio {
...
+            auto domains_iter = domainsbyname.find(domainHash);
+
+            const uint32_t domain_expiration = domains_iter->expiration;
+            const uint32_t present_time = now();
+            fio_400_assert(present_time <= domain_expiration, "fio_address", fa.fioaddress, "FIO Domain expired",
+                           ErrorDomainExpired);
+
+            const uint32_t expiration = fioname_iter->expiration;
+            fio_400_assert(present_time <= expiration, "fio_address", fa.fioaddress, "FIO Address expired.",
+                           ErrorDomainExpired);
```
Comment (lukestokes) [inline contracts/fio.address/fio.address.cpp:1675]: Is there a different error response for an address vs a domain? Is "ErrorDomainExpired" correct?

File: contracts/fio.system/src/voting.cpp [hunk 3]
```diff
@@ -1225,12 +1232,16 @@ namespace eosiosystem {
...
+                    p.fioaddress = fio_address;
+                    p.addresshash = addresshash;
                     p.is_proxy = isproxy;
                     p.is_auto_proxy = false;
                 });
             }else if (!isproxy) { //this is how we undo/clear a proxy
                 name nm;
                 votersbyowner.modify(pitr, same_payer, [&](auto &p) {
+                    p.fioaddress = "";      // TODO: placed here to revert to same state as from regproducer, not sure if correct, verify...
+                    p.addresshash = 0;      // TODO: placed here to revert to same state as from regproducer, not sure if correct, verify...
```
Comment (lukestokes) [inline contracts/fio.system/src/voting.cpp:1244]: We okay with this TODO here?

File: contracts/fio.token/include/fio.token/fio.token.hpp [hunk 4]
```diff
@@ -187,7 +213,7 @@ namespace eosio {
                     if ((numberVestingPayouts == 2)&&
                         ((lockiter->grant_type == 1) ||
                          (lockiter->grant_type == 2) ||
-                         (lockiter->grant_type == 3)))
+                         (lockiter->grant_type == 3)) && doupdate)
```
Comment (lukestokes) [inline contracts/fio.token/include/fio.token/fio.token.hpp:216]: This will fix any broken locked accounts that haven't yet voted, yes?

**Review by adsorptionenthalpy — APPROVED** [review]

**Review by lukestokes — APPROVED** [review]
I don't C, but at first glance seems okay. I think we should do a deep review of these changes as we're testing.

File: contracts/fio.address/fio.address.cpp (hunk 1 above)
Comment (fiocasey) [reply contracts/fio.address/fio.address.cpp:556]: No, not possible with the current implementation. Templated functions are not available on the contract side.

File: contracts/fio.system/src/voting.cpp (hunk 3 above)
Comment (fiocasey) [reply contracts/fio.system/src/voting.cpp:1244]: I don't see the need for it. I placed it in my code cleanup queue.

File: contracts/fio.address/fio.address.cpp (hunk 2 above)
Comment (fiocasey) [reply contracts/fio.address/fio.address.cpp:1675]: This is a common problem in our contracts but the error code has absolutely no effect on the block log or any response. Purely to help a developer manage error codes outsides of the parent contract.

**Pull request merged**